### PR TITLE
frontend/fix/#1840: added check for fakegps at choose city screen for…

### PR DESCRIPTION
**Pull request by ashkriti**
… driver

## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
added check for fakegps at choose city screen

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs [hunk 1]
```diff
@@ -359,6 +370,21 @@ enableLocationPermission state push =
          , margin $ MarginTop 4
          ] <> FontStyle.paragraphText TypoGraphy
   ]
+
+sourceUnserviceableView :: forall w. (Action -> Effect Unit) -> ChooseCityScreenState -> PrestoDOM (Effect Unit) w
```
Comment (MercyQueen) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs:374]: @ashkriti rename this function as it is related to mock location

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs [hunk 2]
```diff
@@ -73,16 +81,18 @@ view push state =
           ) (const AfterRender)
         -- , background if DA.any (_ == state.props.currentStage) [ENABLE_PERMISSION, CAROUSEL, DETECT_LOCATION] then "#FFFAED" else Color.white900
         , gradient (Linear 0.0 ["#F5F8FF", "#E2EAFF"])
-        , padding $ PaddingBottom 24
-        ][ relativeLayout
+        -- , padding $ PaddingBottom 24
+        ][ if ((state.props.isMockLocation && (MU.getMerchant FunctionCall == MU.NAMMAYATRI)) && state.props.currentStage == DETECT_LOCATION) then (sourceUnserviceableView push state) else dummyView
```
Comment (MercyQueen) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs:85]: @ashkriti why do we need a merchant check here?

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/ComponentConfig.purs [hunk 3]
```diff
@@ -97,3 +101,40 @@ getChangeLanguageText value =
...
                     _ -> "Change Language"
     Mb.Nothing -> "Change Language"
+
+sourceUnserviceableConfig :: ChooseCityScreenState -> ErrorModal.Config
+sourceUnserviceableConfig state =
+  let
+    config = ErrorModal.config
+    errorModalConfig' =
+      config
+        { height = if (MU.getMerchant FunctionCall) == MU.NAMMAYATRI && state.props.isMockLocation then MATCH_PARENT else WRAP_CONTENT
```
Comment (MercyQueen) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/ComponentConfig.purs:111]: remove merchant check and use config for this.

**Comment by witcher-shailesh** [comment]
Need more work

File: Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java [hunk 4]
```diff
@@ -675,18 +675,18 @@ public void isMockLocation(String callback) {
...
-                                        callback, isMock);
+                        }
+                        if (callback != null) {   
+                            String js = String.format(Locale.ENGLISH, "window.callUICallback('%s','%s');",
+                                  callback, ( (location != null) ? isMock : "failed"));
                                 bridgeComponents.getJsCallback().addJsToWebView(js);
-                            }
                         }
                     })
                     .addOnFailureListener(activity, e -> Log.e(LOCATION, "Last and current position not known"));
```
Comment (Vignesh-772) [inline Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:692]: add callback for addOnFailureListener also

File: Frontend/ui-driver/src/Flow.purs [hunk 5]
```diff
@@ -2565,7 +2565,10 @@ noInternetScreenFlow triggertype = do
                         false ->  baseAppFlow false Nothing
     TURN_ON_GPS -> if not internetCondition then noInternetScreenFlow "INTERNET_ACTION"
                     else do
-                      (DriverActiveInactiveResp resp) <- Remote.driverActiveInactiveBT "true" $ toUpper $ show Online
+                      when (isTokenValid (getValueToLocalStore REGISTERATION_TOKEN)) $ do 
+                        (DriverActiveInactiveResp resp) <- Remote.driverActiveInactiveBT "true" $ toUpper $ show Online
+                        pure unit 
+                      _ <- pure $ spy "nointernetscreen" "GPS_ACTION"
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Flow.purs:2571]: remove spy

File: Frontend/ui-driver/src/Flow.purs [hunk 6]
```diff
@@ -2565,7 +2565,9 @@ noInternetScreenFlow triggertype = do
                         false ->  baseAppFlow false Nothing
     TURN_ON_GPS -> if not internetCondition then noInternetScreenFlow "INTERNET_ACTION"
                     else do
-                      (DriverActiveInactiveResp resp) <- Remote.driverActiveInactiveBT "true" $ toUpper $ show Online
+                      when (isTokenValid (getValueToLocalStore REGISTERATION_TOKEN)) $ do 
+                        (DriverActiveInactiveResp resp) <- Remote.driverActiveInactiveBT "true" $ toUpper $ show Online
+                        pure unit
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Flow.purs:2570]: can you update it as `void $  Remote.driverActiveInactiveBT "true" $ toUpper $ show Online` no need of `pure unit`

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/ComponentConfig.purs [hunk 7]
```diff
@@ -97,3 +97,35 @@ getChangeLanguageText value =
                     "Bangalore" -> "ಭಾಷೆ ಬದಲಾಯಿಸಿ"
                     _ -> "Change Language"
     Mb.Nothing -> "Change Language"
+
+mockLocationConfig :: ChooseCityScreenState -> ErrorModal.Config
+mockLocationConfig state =
+  let
+    config = ErrorModal.config
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/ComponentConfig.purs:104]: Update like this

```
mockLocationConfig :: ChooseCityScreenState -> ErrorModal.Config
mockLocationConfig state =
  ErrorModal.config
        { height = MATCH_PARENT
        , background = Color.white900
        , corners = Corners 24.0 true true false false
        , stroke = ("1," <> Color.borderGreyColor)
        , imageConfig
          { imageUrl = fetchImage FF_ASSET "ny_ic_location_unserviceable"
          , height = V 99
          , width = V 133
          , margin = MarginVertical 50 20
          }
        , errorConfig
          { text = getString UNABLE_TO_GET_YOUR_LOCATION
          , color = Color.black800
          , margin = MarginBottom 5
          }
        , errorDescriptionConfig
          { text = getString TURN_OFF_ANY_MOCK_LOCATION_APP_AND_RESTART
          , color = Color.black700
          , margin = Margin 20 0 20 40
          }
        , buttonConfig
          { visibility = GONE }
        }
```

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/Controller.purs [hunk 8]
```diff
@@ -127,6 +130,10 @@ eval UpdateLocationPermissionState state = do
   let newState = state {props {isLocationPermissionGiven = true, currentStage = DETECT_LOCATION}}
   updateAndExit newState $ RefreshScreen newState
 
+eval (IsMockLocation isMock) state = 
+      -- _ = unsafePerformEffect $ if val then  logEvent (state.data.logField) "ny_fakeGPS_enabled" else pure unit -- we are using unsafePerformEffect becasue without it we are not getting logs in firebase, since we are passing a parameter from state i.e. logField then the output will be inline and it will not be able to precompute so it's safe to use it here.
+  continue state{props{isMockLocation = (isMock == "true") },data{ locationDetectionFailed = (isMock == "failed")}}
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/Controller.purs:135]: remove extra brackets

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/Controller.purs [hunk 9]
```diff
@@ -127,6 +130,10 @@ eval UpdateLocationPermissionState state = do
   let newState = state {props {isLocationPermissionGiven = true, currentStage = DETECT_LOCATION}}
   updateAndExit newState $ RefreshScreen newState
 
+eval (IsMockLocation isMock) state = 
+      -- _ = unsafePerformEffect $ if val then  logEvent (state.data.logField) "ny_fakeGPS_enabled" else pure unit -- we are using unsafePerformEffect becasue without it we are not getting logs in firebase, since we are passing a parameter from state i.e. logField then the output will be inline and it will not be able to precompute so it's safe to use it here.
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/Controller.purs:134]: Why local is removed ?

**Review by Vignesh-772 — APPROVED** [review]

File: Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java [hunk 10]
```diff
@@ -672,24 +672,32 @@ public void getCurrentPosition(String callback) {
     public void isMockLocation(String callback) {
         Activity activity = bridgeComponents.getActivity();
         if (!isLocationPermissionEnabled()) return;
-        if (client != null && activity != null)
+        if (client != null && activity != null) {
             client.getLastLocation()
                     .addOnSuccessListener(activity, location -> {
-                        boolean isMock;
+                        boolean isMock = false;
```
Comment (witcher-shailesh) [inline Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:678]: in java, default value of a boolean is false, no need to explicitly mention it as `false`

File: Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java [hunk 11]
```diff
@@ -672,24 +672,32 @@ public void getCurrentPosition(String callback) {
...
                             } else {
                                 isMock = location.isMock();
                             }
-                            if (callback != null) {
-                                String js = String.format(Locale.ENGLISH, "window.callUICallback('%s','%s');",
-                                        callback, isMock);
+                        }
+                        if (callback != null) {
+                            String js = String.format(Locale.ENGLISH, "window.callUICallback('%s','%s');",
+                                  callback, ((location != null) ? isMock : "failed"));
```
Comment (witcher-shailesh) [inline Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:688]: failed can be false, don't make it unsafe as isMock is a boolean and in else condition we are sending a string. Failed can be treated as false

File: Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java [hunk 12]
```diff
@@ -672,24 +672,32 @@ public void getCurrentPosition(String callback) {
...
                                 bridgeComponents.getJsCallback().addJsToWebView(js);
-                            }
                         }
                     })
-                    .addOnFailureListener(activity, e -> Log.e(LOCATION, "Last and current position not known"));
+                    .addOnFailureListener(activity, e -> {
+                        Log.e(LOCATION, "Last and current position not known");
+                        if (callback != null) {
+                            String js = String.format(Locale.ENGLISH, "window.callUICallback('%s','%s');",
+                                    callback, "failed");
```
Comment (witcher-shailesh) [inline Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:696]: replace "failed" with `false`

File: Frontend/ui-driver/src/MerchantConfigs/DefaultConfig.purs [hunk 13]
```diff
@@ -163,5 +163,5 @@ config =
               supportNumber : ""
             }
         ]
-  , enableMockLocation : false
+  , enableMockLocation : true
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/MerchantConfigs/DefaultConfig.purs:166]: default can be false

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs [hunk 14]
```diff
@@ -24,29 +23,35 @@ import JBridge as JB
...
   , name: "ChooseCity"
   , globalEvents:  [(\ push -> do    
-    if not initialState.props.isLocationPermissionGiven then do
-      _ <- JB.storeCallBackDriverLocationPermission push LocationPermissionCallBack
-      pure unit 
+    if not initialState.props.isLocationPermissionGiven then do 
+      JB.storeCallBackDriverLocationPermission push LocationPermissionCallBack
+      pure unit
     else pure unit
+    when (initialState.data.config.enableMockLocation) $
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs:48]: these when function are working here? did you confirmed it? @ashkriti

File: Frontend/ui-driver/src/Screens/HomeScreen/View.purs [hunk 15]
```diff
@@ -95,7 +95,7 @@ screen initialState =
         ( \push -> do
           _ <- pure $ JB.checkAndAskNotificationPermission false
           _ <- pure $ printLog "initial State" initialState
-          _ <- if initialState.data.config.enableMockLocation then JB.isMockLocation push IsMockLocation else pure unit
+          -- _ <- if initialState.data.config.enableMockLocation then JB.isMockLocation push IsMockLocation else pure unit
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/HomeScreen/View.purs:98]: please add a comment as well along with the commented code, what is this commented code about

**Review by witcher-shailesh — CHANGES_REQUESTED** [review]

File: Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java (hunk 10 above)
Comment (ashkriti) [reply Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:678]: the isMock variable was being assigned within an if (location != null) block that is why need to initialise the variable with a default value.

Comment (witcher-shailesh) [reply Frontend/android-native/mobility-common/src/main/java/in/juspay/mobility/common/MobilityCommonBridge.java:678]: this is not explanatory enough, still didn't get why is this required

File: Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs (hunk 14 above)
Comment (ashkriti) [reply Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs:48]: yes it is working fine

Comment (witcher-shailesh) [reply Frontend/ui-driver/src/Screens/ChooseCityScreen/View.purs:48]: Last time when I checked this config was not working

**Review by witcher-shailesh — APPROVED** [review]

**Pull request merged**